ports/nrf/common-hal/pulseio/PulseOut.c

```diff
@@ -29,31 +29,136 @@
+// the code is single threaded.
+static uint8_t refcount = 0;
 
-void pulseout_reset() {
+static nrfx_timer_t *timer = NULL;
+
+static uint16_t *pulse_array = NULL;
+static volatile uint16_t pulse_array_index = 0;
+static uint16_t pulse_array_length;
+
+static void turn_on(pulseio_pulseout_obj_t *pulseout) {
+    pulseout->pwmout->pwm->PSEL.OUT[0] = pulseout->pwmout->pin_number;
+}
+
+static void turn_off(pulseio_pulseout_obj_t *pulseout) {
+    // Disconnect pin from PWM.
+    pulseout->pwmout->pwm->PSEL.OUT[0] = 0xffffffff;
+    // Make sure pin is low.
+    nrf_gpio_pin_clear(pulseout->pwmout->pin_number);
+}
+
+static void start_timer(void) {
+    nrfx_timer_clear(timer);
+    // true enables interrupt.
+    nrfx_timer_compare(timer, NRF_TIMER_CC_CHANNEL0, pulse_array[pulse_array_index], true);
+    nrfx_timer_resume(timer);
+}
+
+static void pulseout_event_handler(nrf_timer_event_t event_type, void *p_context) {
+    pulseio_pulseout_obj_t *pulseout = (pulseio_pulseout_obj_t*) p_context;
+    if (event_type != NRF_TIMER_EVENT_COMPARE0) {
+        // Spurious event.
+        return;
+    }
+    nrfx_timer_pause(timer);
+
+    pulse_array_index++;
 
+    // No more pulses. Turn off output and don't restart.
+    if (pulse_array_index >= pulse_array_length) {
+        turn_off(pulseout);
+        return;
+    }
+
+    // Alternate on and off, starting with on.
+    if (pulse_array_index % 2 == 0) {
+        turn_on(pulseout);
+    } else {
+        turn_off(pulseout);
+    }
+
+    // Count up to the next given value.
+    start_timer();
 }
 
-void common_hal_pulseio_pulseout_construct(pulseio_pulseout_obj_t* self, const pulseio_pwmout_obj_t* carrier) {
-    mp_raise_NotImplementedError(NULL);
+void pulseout_reset() {
+    if (timer != NULL) {
+        nrf_peripherals_free_timer(timer);
+    }
+    refcount = 0;
+}
+
+void common_hal_pulseio_pulseout_construct(pulseio_pulseout_obj_t* self,
+                                           const pulseio_pwmout_obj_t* carrier) {
+    if (refcount == 0) {
+        timer = nrf_peripherals_allocate_timer();
+        if (timer == NULL) {
+            mp_raise_RuntimeError(translate("All timers in use"));
+        }
+    }
+    refcount++;
+
+    nrfx_timer_config_t timer_config = {
+        // PulseOut durations are in microseconds, so this is convenient.
+        .frequency = NRF_TIMER_FREQ_1MHz,
+        .mode = NRF_TIMER_MODE_TIMER,
+        .bit_width = NRF_TIMER_BIT_WIDTH_32,
+        .interrupt_priority = NRFX_TIMER_DEFAULT_CONFIG_IRQ_PRIORITY,
+        .p_context = self,
+    };
+
+    self->pwmout = carrier;
+
+    nrfx_timer_init(timer, &timer_config, &pulseout_event_handler);
+    turn_off(self);
 }
 
 bool common_hal_pulseio_pulseout_deinited(pulseio_pulseout_obj_t* self) {
-  return 1;
+    return self->pwmout == NULL;
 }
 
 void common_hal_pulseio_pulseout_deinit(pulseio_pulseout_obj_t* self) {
+    if (common_hal_pulseio_pulseout_deinited(self)) {
+        return;
+    }
+    turn_on(self);
```

Comment:
```suggestion
    turn_off(self);
```

Should this turn off instead?

Reply:
This is the same behavior as in the `atmel-samd` port. When you deinit, it turns the carrier pin back on, as it was before you wrapped it in the `PulseOut()`.

Reply:
https://github.com/adafruit/circuitpython/blob/master/ports/atmel-samd/common-hal/pulseio/PulseOut.c#L166

Reply:
Ah ok! Thanks for the reminder.